packages/backend/src/functions/participant.ts

```diff
@@ -0,0 +1,477 @@
+import * as functions from "firebase-functions"
+import admin from "firebase-admin"
+import dotenv from "dotenv"
+import { commonTerms, getCircuitsCollectionPath, getParticipantsCollectionPath } from "@zkmpc/actions/src"
+import { CeremonyState, ParticipantStatus, ParticipantContributionStep } from "@zkmpc/actions/src/types/enums"
+import { ParticipantDocument } from "@zkmpc/actions/src/types"
+import {
+    PermanentlyStoreCurrentContributionTimeAndHash,
+    TemporaryStoreCurrentContributionMultiPartUploadId,
+    TemporaryStoreCurrentContributionUploadedChunkData
+} from "types"
+import {
+    getCeremonyCircuits,
+    getCurrentServerTimestampInMillis,
+    getDocumentById,
+    queryNotExpiredTimeouts
+} from "../lib/utils"
+import { COMMON_ERRORS, logAndThrowError, printLog, SPECIFIC_ERRORS } from "../lib/errors"
+import { LogLevel } from "../../types/enums"
+
+dotenv.config()
+
+/**
+ * Check the user's current participant status for the ceremony.
+ * @notice this cloud function has several tasks:
+ * 1) Check if the authenticated user is a participant
+ * 1.A) If not, register it has new participant for the ceremony.
+ * 1.B) Otherwise:
+ * 2.A) Check if already contributed to all circuits or,
+ * 3.A) If already contributed, return false
+ * 2.B) Check if it has a timeout in progress
+ * 3.B) If timeout expired, allows the participant to resume the contribution.
+ * 3.C) Otherwise, return false.
+ * 1.D) If no timeout / participant already exist, just return true.
+ * @dev true when the participant can participate (1.A, 3.B, 1.D); otherwise false.
+ */
+export const checkParticipantForCeremony = functions.https.onCall(
+    async (data: { ceremonyId: string }, context: functions.https.CallableContext) => {
+        if (!context.auth || (!context.auth.token.participant && !context.auth.token.coordinator))
+            logAndThrowError(SPECIFIC_ERRORS.SE_AUTH_NO_CURRENT_AUTH_USER)
+
+        if (!data.ceremonyId) logAndThrowError(COMMON_ERRORS.CM_MISSING_OR_WRONG_INPUT_DATA)
+
+        // Prepare Firestore DB.
+        const firestore = admin.firestore()
+
+        // Get data.
+        const { ceremonyId } = data
+        const userId = context.auth?.uid
+
+        // Look for the ceremony document.
+        const ceremonyDoc = await getDocumentById(commonTerms.collections.ceremonies.name, ceremonyId)
```

Comment:
You usually do a check like: if (!ceremonyDoc.exists) throw - should this be added here?

Reply:
`getDocumentById` itself includes the document existence check!